specification/dartLangSpec.tex

```diff
@@ -2531,11 +2535,22 @@ \section{Classes}
 (\ref{theMethodNoSuchMethod}).
 It is a compile-time error if $C$ has
 a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is not a correct override of $m$
+with a method signature \DefineSymbol{m'}
+which is not a correct override of $m$
```

Comment:
So if `m'` is `int Function(covariant int)` and `m` is `int Function(num)`, then `m'` is a correct override of `m`?
But `m'` as `int Function(int)` is not a correct override of `m` as `int Function(covariant num)`?

(I think this is probably correct for what we currently intended to specify, but I do want both to be invalid. Eventually :).

Reply:
First a few remarks on the newest upload:

I added a few words at 'must fully implement', to make it explicit that this rule is about inherited methods, not about methods that are declared in `$C$`. That should be a benign change (because the declarations in `$C$` must satisfy the 'correct override' constraint as well, but that's in a different section).

Next, I moved the new paragraph down below the commentary, in order to avoid some confusion about the topic of that commentary.

I added a few words to the new paragraph (introducing `m'` afresh, because it's now further away). I think that's clearer.

Now, about the actual rule. I think it says what you want, and also that it was possible to get this effect without too many twisted turns. So that's nice!

The question about whether any given parameter in the inherited signature `m'` is cov-by-decl is answered by checking `m'` itself. That is, a parameter inherited from `B` is _not_ cov-by-decl at this point if it isn't known to be cov-by-decl in `B`.

This ensures that we get an error for "magically becoming cov-by-decl when inherited" (using the rule in Correct Member Overrides), e.g.

```dart
class A { m(num n) {}}
abstract class I { m(covariant num n); }
class B extends A implements I {} // Error.
```

> So if `m'` is `int Function(covariant int)` and `m` is `int Function(num)`, 
> then `m'` is a correct override of `m`?

I think we couldn't actually have those member signatures:

`m'` is the inherited signature (the spec writes that as `int m(covariant int)` because we don't support `covariant` in function types) and `m` is the one from the interface (`int m(num)`).

But if the parameter is cov-by-decl in any superclass then it would also be cov-by-decl in the interface of the current class.

But if the superclass has signature `int m(int)` and the class interface has `int m(covariant num)` then we get a compile-time error: `int m(int)` is not a correct override. We don't "magically" add `covariant` to the inherited signature, and it's a violation of the cov-by-decl rule in 'Correct Member Overrides' when `covariant` is missing.

Consider the opposite situation:

```dart
class A { m(covariant int i) {}}
abstract class I { m(num n); }
class B extends A implements I {} // OK.
```

This is actually accepted: The inherited signature is `dynamic m(covariant int)` and that is a correct override of `dynamic m(num)`.

It seems rather reasonable to me, but do you want this to be an error?

> But `m'` as `int Function(int)` is not a correct override of `m` as `int Function(covariant num)`?

So this would correspond to the following code:

```dart
class A { int m(int i) {}}
abstract class I { int m(covariant num n); }
class B extends A implements I {}
```

Yes, that would be an error, because `int m(int)` is not a correct override of `int m(covariant num)`: `covariant` is missing (and more, but that's enough to make it an error).

Reply:
I think this is fine, and we'll have to further discuss whether
```dart
class A { m(covariant int i) {}} 
abstract class I { m(num n); } 
class B extends A implements I {} // OK.
```
should be an error too (which I think it should because `B().m(2.2)` is known to throw a type error, meaning that the class doesn't implement it *own* interface, without even needing to up-cast it, so it's not a *covariance* based error, it's just a plain error).